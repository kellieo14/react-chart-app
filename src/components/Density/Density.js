import React, { useState, useEffect } from 'react';
import Polar from '../Graphs/Polar/Polar';
import {sort} from '../../Utils';

export const extractDensity = (countries, densityCount) => {
    for (let country in countries) {
        if (countries[country].area === 0 || countries[country].population === 0 || countries[country].area === null) {
            densityCount.push([countries[country].name, 0]);
        } else {
            let density = (countries[country].name, (countries[country].population) / (countries[country].area))
            densityCount.push([countries[country].name, Math.round(density)]);
        }
    }
}

export const sortCountriesByHighestDensity = (densityCount, labels, densityArr) => {
    sort(densityCount, 'con', 5, labels, densityArr);
}

export const setTopDensestCountries = (labels, setData, densityArr) => {
    if (labels.length > 0) {
        setData({
            title: 'Countries with Highest Density per Square Kilometer',
            position: 'right',
            labels: labels,
            datasets: [
                {
                    label: 'People per Square Kilometer',
                    backgroundColor: ['rgb(11,226,215, .8)','rgb(167,62,92, .8)', 'rgb(0,128,214, .8)','rgb(141,198,63, .8)', 'rgb(255,225,25, .8)'],
                    borderWidth: 1,
                    hoverBackgroundColor: ['rgb(11,226,215)','rgb(167,62,92)', 'rgb(0,128,214)','rgb(141,198,63)', 'rgb(255,225,25)'],
                    data: densityArr,
                }
            ]
        })
    }
}

const Density = (props) => {
    const [data, setData] = useState({});
    
    useEffect(() => {
        let densityCount = [];
        let labels = [];
        let densityArr = [];

        extractDensity(props.countries, densityCount);
        sortCountriesByHighestDensity(densityCount, labels, densityArr);
        setTopDensestCountries(labels, setData, densityArr);
    }, [props.countries]);

    return (
        <div>
            <Polar data={data} />
        </div>
    );
}

export default Density;