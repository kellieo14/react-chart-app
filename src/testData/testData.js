export const countries = [
    {
        name: 'China', 
        region: 'Asia', 
        languages: [
            { iso639_1: "zh", iso639_2: "zho", name: "Chinese", nativeName: "中文 (Zhōngwén)"}
        ], 
        population: 1377422166,
        area: 9640011,
    }, 
    {
        name: 'India', 
        region: 'Asia',
        languages: [
            {iso639_1: "hi", iso639_2: "hin", name: "Hindi", nativeName: "हिन्दी"},
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 1295210000,
        area: 3287590,
    }, 
    {
        name: 'United States of America', 
        region: 'Americas', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 323947000,
        area: 9629091,
    },
    {
        name: 'Nigeria', 
        region: 'Africa', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 186988000,
        area: 923768,
    }, 
    {
        name: 'France', 
        region: 'Europe', 
        languages: [
            {iso639_1: "fr", iso639_2: "fra", name: "French", nativeName: "français"}
        ], 
        population: 66710000,
        area: 640679,
    }, 
    {
        name: 'Australia', 
        region: 'Oceania', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 24117360,
        area: 7692024,
    }, {
        name: 'Polarland', 
        region: 'Polar', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 11239004,
        area: 1284000,
    }, {
        name: 'Otherland', 
        region: 'Other', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 100,
        area: 619745,
    }
];