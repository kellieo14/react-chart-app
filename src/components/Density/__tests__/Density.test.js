import React from 'react';
import * as reactHooks from 'react';
import TestRenderer from 'react-test-renderer';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { countries } from '../../../testData/testData';
import Density, { extractDensity, sortCountriesByHighestDensity } from '../Density';

Enzyme.configure({adapter: new Adapter()});

describe('Density', () => {
    let spyUseState;
    let spyUseEffect;
    let mockSetData;

    beforeEach(() => {
        mockSetData = jest.fn();
        spyUseState = jest.spyOn(reactHooks, 'useState')
            .mockImplementationOnce(() => [{}, mockSetData]);
        spyUseEffect = jest.spyOn(reactHooks, 'useEffect');
        
    })
    const expectedDensityCount = [
        ['China', 143],
        ['India', 394],
        ['United States of America', 34], 
        ['Nigeria', 202], 
        ['France', 104],
        ['Australia', 3],
        ['Polarland', 9],
        ['Otherland', 0],
    ]

    const expectedDensityArr = [394, 202, 143, 104, 34];
    const expectedLabels = ['India', 'Nigeria', 'China', 'France', 'United States of America'];

    it('should match snapshot', () => {
        const testRenderer = TestRenderer.create(<Density />)
        expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    it('should call useState and useEffect', () => {
        mount(<Density countries={countries} />);
        expect(spyUseState).toHaveBeenCalled();
        expect(spyUseEffect).toHaveBeenCalled();
    })

    it('should get density count', () => {
        let densityCount = [];
        extractDensity(countries, densityCount);
        expect(densityCount).toEqual(expectedDensityCount);
    });

    it('should sort countries by highest density', () => {
        let labels = [];
        let densityArr =[];
        sortCountriesByHighestDensity(expectedDensityCount, labels, densityArr);
        expect(densityArr).toEqual(expectedDensityArr);
        expect(labels).toEqual(expectedLabels);
    });
});