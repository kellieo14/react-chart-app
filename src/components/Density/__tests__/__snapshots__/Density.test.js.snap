// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Density should match snapshot 1`] = `
<div>
  <div
    className="polar"
  />
</div>
`;
